### DISCUSSION TITLE
How can I launch qutebrowser for debugging from VSCode?

### DISCUSSION THREAD

**Discussion by Aster89** [qutebrowser/qutebrowser#7417] [Q&A]
I'm not really using VSCode, as I use [vimspector](https://puremourning.github.io/vimspector/) on Vim, but in order to reach a wider audience, I've reproduced the issue in VSCode as well.

The following `launch.json` config file is enough for the play button of VSCode to launch the chrome/chromium browser in debug mode. I can then step through the JavaScript code of the web page I'm serving on the `"url"` by executing `php -S localhost:1234` from the direcotry in which `index.html` is.
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "chrome",
            "request": "launch",
            "name": "Launch Chrome against localhost",
            "url": "http://localhost:1234",
            "webRoot": "${workspaceFolder}"
        }
    ]
}
```

I've tried adapting it for qutebrowser like this,
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "chrome",
            "request": "launch",
            "runtimeExecutable": "/usr/bin/qutebrowser",
            "runtimeArgs": ["-T", "--qt-flag remote-debugging-port=9229"],
            "name": "Launch Chrome against localhost",
            "url": "http://localhost:1234",
            "webRoot": "${workspaceFolder}"
        }
    ]
}
```
where I've given the arguments according to the answer/comment I received to [a previous question of mine](https://github.com/qutebrowser/qutebrowser/discussions/7152), but this doesn't work, as all I get is an _Unable to attach to browser_ error dialog.

**Comment by The-Compiler** [comment]
`"--qt-flag remote-debugging-port=9229"` should be passed as two separate arguments, rather than one argument with spaces. Hoping that'll work better!

**Reply by Aster89** [reply]
Still getting the same error :/

**Reply by The-Compiler** [reply]
I can reproduce... I also tried `"env": {"QTWEBENGINE_REMOTE_DEBUGGING": "9229"},`, but no success with that either.

I'm out of ideas I'm afraid. Would be good to get some more information from VS Code about what's wrong (given that it doesn't seem to launch qutebrowser at all!), but I was unable to find out anything.

**Reply by Aster89** [reply]
Shall we file something on [VSCode](https://github.com/microsoft/vscode/issues)? I mean, maybe the problem is on qutebrowser side, but VSCode is not really doing anything to help in terms of error messages. But I'm unsure about the info I should provide.

**Reply by The-Compiler** [reply]
Feel free! I've never used this VS Code feature before, so I don't know what the proper place to look at for error messages would be. I tried to go through the different channels in the Output pane, but didn't seem to find something.

**Reply by Aster89** [reply]
Done: https://github.com/microsoft/vscode/issues/162957.

I foresee that we'll get nothing out of it :smile:

Unless you have a better experience with VSCode :thinking:

**Reply by Aster89** [reply]
At the moment, with this confi
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Qutebrowser - Launch against localhost",
            "type": "chrome",
            "request": "launch",
            "trace": true,
            "includeLaunchArgs": false,
            "runtimeExecutable": "/usr/bin/qutebrowser",
            "runtimeArgs": ["-T", "--qt-flag remote-debugging-port=9229"],
            "url": "http://localhost:1234",
            "webRoot": "${workspaceFolder}"
        }
    ]
}
```
is that VSCode does launch qutebrowser in a new window (so I think `-T` is doing its job?), but it opens the web search for the string `--qt-flag remote-debugging-port=9229`, rather than going to the `localhost:1234`, which I've serving via `php -S localhost:1234 > /dev/null 2>&1 &`; I can even go manually there and I do see the page, but VSCode keeps waiting for something (I see the little bar looping in the top left, in the Run and Debug panel), with this being in the DEBUG CONSOLE:
```
Note: Using the "preview" debug extension
Verbose logs are written to:
/home/enrico/.config/Code - OSS/logs/20230426T105552/window1/exthost/ms-vscode.js-debug-nightly/vscode-debugadapter-cedb0444.json.gz
```

Do you have any more clues with this info?

**Reply by The-Compiler** [reply]
`--qt-flag remote-debugging-port=9229` should be two arguments, not one.

**Reply by Aster89** [reply]
Thanks. However, the browser still doesn't go to the `url` I've configured in the JSON.

In the console, I get this additional line:
```
DevTools listening on ws://127.0.0.1:9229/devtools/browser/c1f58a87-6d77-495a-ba11-636878124b6b
```
Does this help me?
